### PR TITLE
frozen-abi: extend support for collections in StableAbi

### PR DESCRIPTION
**Pull request by puhtaytow**
### Description

Collections in StableAbi were supported thru sampling expression using `stable_abi_sample(with = "<code>")` and thru manual implementation of rand distribution, which demanded highly verbose treatment in every case.

### Summary of changes
- added support for Vec, VecDeque, HashMap, HashSet, BtreeMap, BtreeSet
- added two context types SequenceLenRange and SequenceLenMax
- added new macro impl_with_context_via
- added test macros, to generate equivalent test types for collections, with split for multiple and single element
- updated docs

### CODE REVIEW

File: frozen-abi/src/stable_abi/impls.rs [hunk 1]
```diff
@@ -43,6 +50,37 @@ macro_rules! impl_stable_abi_for_tuples {
     };
 }
 
+macro_rules! impl_stable_abi_via_default_ctx {
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: I suspect we could remove this macro and support default case with impl_with_context_via (even now with something like `() as _def_ctx => MyDefaultContext`, but probably it could be made more ergonomic by allowing extra syntax like `() => MyDefaultContext` or `default = MyDefaultContext`

File: frozen-abi/src/stable_abi/impls.rs [hunk 2]
```diff
@@ -43,6 +50,37 @@ macro_rules! impl_stable_abi_for_tuples {
...
+    ) => {
+        impl<$($g)*> StableAbi for $self_ty where $($wc)* {
+            fn random_with_context(rng: &mut (impl RngCore + ?Sized), _ctx: ()) -> Self {
+                Self::random_with_context(rng, $default)
+            }
+        }
+    };
+}
+
+macro_rules! impl_with_context_via {
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: let's put the macro into context.rs

File: frozen-abi/src/stable_abi/impls.rs [hunk 3]
```diff
@@ -504,4 +681,70 @@ mod tests {
...
+        derive(
+            solana_frozen_abi_macro::StableAbi,
+            solana_frozen_abi_macro::StableAbiSample
+        ),
+        solana_frozen_abi_macro::frozen_abi(
+            abi_digest = "9SaZmoaudvkdtR62vkPuf1RtzEJ8sbqN4LkUJcXGesDL",
+            abi_serializer = "wincode",
+        )
+    )]
+    struct TestStableAbiSampleSizeAttribute {
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: instead of adding this one kitchen sink struct, which only tests support for syntax, let's add actual compatibility tests for supported equivalences, e.g.
```rust
struct MultiElementsWith {
   #[stable_abi_sample(with = "..")
   elements: Vec<i32>
}
struct MultiElementsVec {
   #[stable_abi_sample(ctx = SequenceLenMax(5))
   elements: Vec<i32>
}
struct MultiElementsVecDefault {
   elements: Vec<i32>
}
struct MultiElementsVecDeque {
   #[stable_abi_sample(ctx = SequenceLenMax(5))
   elements: VecDeque<i32>
}


struct UpToOneElementWith {
   #[stable_abi_sample(with = "..")
   elements: Vec<i32>
}
struct UpToOneElementVec {
   #[stable_abi_sample(ctx = SequenceLenMax(1))
   elements: Vec<i32>
}
struct UpToOneElementBTreeSet {
   #[stable_abi_sample(ctx = SequenceLenMax(1))
   elements: BTreeSet<i32>
}
```

each kind will have the same ABI digest and I suppose we will have many such structs, might make sesne to generate them with a macro

File: frozen-abi/src/stable_abi/impls.rs [hunk 4]
```diff
@@ -104,11 +142,150 @@ where
...
+
+impl_with_context_via! {
+    impl[T, S] for HashSet<T, S>,
+    SequenceLenMax as ctx => SequenceLenRange { min: 0, max: ctx.0 },
+    where { T: StableAbi + Eq + Hash, S: BuildHasher + Default },
+}
+
+impl_stable_abi_via_default_ctx! {
+    impl[T] for Vec<T>,
+    default = SequenceLenRange { min: 0, max: 3 },
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: 3 seems fine, though maybe let's use slightly larger value and keep it non-power-of-2, maybe 5?

File: frozen-abi/src/stable_abi/mod.rs [hunk 5]
```diff
@@ -1,5 +1,6 @@
 use rand::RngCore;
 
+mod context;
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/mod.rs]: should be `pub`

File: frozen-abi/src/stable_abi/impls.rs (hunk 3 above)
Comment (puhtaytow) [reply frozen-abi/src/stable_abi/impls.rs]: Should be ok now*, do we need more granularity, cases?

File: frozen-abi/src/stable_abi/impls.rs [hunk 6]
```diff
@@ -504,4 +652,231 @@ mod tests {
...
+        multiple_elements = "9QBF2ReRkBJ4ssnkrAcWq6Cb1tghYt55MLZzZQnuo7Zz",
+        up_to_one_element = "3cFop5HjPdJhyL23aaE6tzdnXiintAmgbBHgHLoGvNS9",
+        fields = [a: u8, b: bool, c: char, d: i16, e: u128,]
+    );
+    impl_equivalent_sequence_sample_structs!(
+        module = equivalent_sequence_sample_single,
+        multiple_elements = "762V1KB6NRDroyHo31bjg7gwX9nx8QqHCP9EJ3Y7nLGE",
+        up_to_one_element = "7dXyMka1Z72sENE9vva8vmE65F7y6kXZKQ2DHu9aTWyz",
+        fields = [a: u8]
+    );
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs:841]: I was thinking more about something like
```rust
impl_sequence_sample_structs!(
    module = equivalent_sequence_sample_single,
    digests = [
        "762V1KB6NRDroyHo31bjg7gwX9nx8QqHCP9EJ3Y7nLGE" => [
            struct MultiElementsWith {
                #[stable_abi_sample(with = "(0..rng.random_range(0..=5)).map(|_| rng.random()).collect()")]
                a: Vec<u8>,
            },
            struct MultiElementsVec {
                #[stable_abi_sample(ctx = SequenceLenMax(5))]
                a: Vec<u8>,
            },
            struct MultiElementsVecDefault {
                a: Vec<u8>,
            },
            struct MultiElementsVecDeque {
                #[stable_abi_sample(ctx = SequenceLenMax(5))]
                a: VecDeque<u8>,
            },
        ],
        "7dXyMka1Z72sENE9vva8vmE65F7y6kXZKQ2DHu9aTWyz" => [
            struct UpToOneElementWith {
                #[stable_abi_sample(with = "(0..rng.random_range(0..=1)).map(|_| rng.random()).collect()")]
                a: Vec<u8>,
            },
            struct UpToOneElementVec {
                #[stable_abi_sample(ctx = SequenceLenMax(1))]
                a: Vec<u8>,
            },
            struct UpToOneElementBTreeSet {
                #[stable_abi_sample(ctx = SequenceLenMax(1))]
                a: BTreeSet<u8>,
            },
        ]
    ]
);
```

impl of the macro:
```rust
macro_rules! impl_sequence_sample_struct {
    (
        derives = ($($derive:path),+),
        serializer = $abi_serializer:literal,
        abi_digest = $abi_digest:expr,
        struct $struct_name:ident { $($body:tt)* }
    ) => {
        #[derive($($derive),+)]
        #[cfg_attr(
            feature = "frozen-abi",
            derive(
                StableAbi, StableAbiSample
            ),
            solana_frozen_abi_macro::frozen_abi(
                abi_digest = $abi_digest,
                abi_serializer = $abi_serializer,
            )
        )]
        struct $struct_name { $($body)* }
    };
}

macro_rules! impl_sequence_sample_structs_for_serializer {
    (
        module = $module_name:ident,
        serializer = $abi_serializer:literal,
        derives = ($($derive:path),+ $(,)?),
        digests = [$(
            $abi_digest:expr => [$(
                struct $struct_name:ident { $($body:tt)* }
            ),+ $(,)?]
        ),+ $(,)?]
    ) => {
        mod $module_name {
            $($(
                impl_sequence_sample_struct!(
                    derives = ($($derive),+),
                    serializer = $abi_serializer,
                    abi_digest = $abi_digest,
                    struct $struct_name { $($body)* }
                );
            )+)+
        }
    };
}

macro_rules! impl_sequence_sample_structs {
    (
        module = $module_name:ident,
        digests = [$(
            $abi_digest:expr => [$(
                struct $struct_name:ident { $($body:tt)* }
            ),+ $(,)?]
        ),+ $(,)?]
    ) => {
        mod $module_name {
            impl_sequence_sample_structs_for_serializer!(
                module = wincode,
                serializer = "wincode",
                derives = (SchemaWrite),
                digests = [$(
                    $abi_digest => [$(
                        struct $struct_name { $($body)* }
                    ),+]
                ),+]
            );
            impl_sequence_sample_structs_for_serializer!(
                module = bincode,
                serializer = "bincode",
                derives = (Serialize),
                digests = [$(
                    $abi_digest => [$(
                        struct $struct_name { $($body)* }
                    ),+]
                ),+]
            );
        }
    };
}
```
```

Comment (kskalski) [reply frozen-abi/src/stable_abi/impls.rs:841]: this way we can add a few more tests:
* similar as above, but with different item type, e.g. `(u8, i32)`
* nested collection `Vec<Vec<>>` 
* map type - equivalence of BTreeMap<K, V> with Vec<(K, V)> and Vec<KeyValue>
* context where min len is set

File: frozen-abi/src/stable_abi/impls.rs [hunk 7]
```diff
@@ -504,4 +652,231 @@ mod tests {
...
+                        )]
+                        $field_name: Vec<$field_ty>,
+                    )+
+                }
+
+                #[derive($($derive),+)]
+                #[cfg_attr(
+                    feature = "frozen-abi",
+                    derive(
+                        solana_frozen_abi_macro::StableAbi,
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: let's just add all necessary imports in the tests module, so all that code is shorter

File: frozen-abi/src/stable_abi/impls.rs (hunk 6 above)
Comment (kskalski) [reply frozen-abi/src/stable_abi/impls.rs:841]: for 1-element collections, we can actually also add hash-based collections

Comment (puhtaytow) [reply frozen-abi/src/stable_abi/impls.rs:841]: Yeah, looks better thanks 👍

File: frozen-abi/src/stable_abi/impls.rs [hunk 8]
```diff
@@ -504,4 +653,189 @@ mod tests {
...
+    macro_rules! impl_sequence_sample_structs {
+    (
+        digests = [$(
+            $abi_digest:expr => [$(
+                struct $struct_name:ident { $($body:tt)* }
+            ),+ $(,)?]
+        ),+ $(,)?]
+    ) => {
+        mod sequence_sample_test_types {
+            impl_sequence_sample_structs_for_serializer!(
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: looking at it again, I think it's actually cleaner to inline this macro, so it would just be:
```rust
macro_rules! impl_sequence_sample_structs {
    (
        module = $module_name:ident,
        digests = [$(
            $abi_digest:expr => [$(
                struct $struct_name:ident { $($body:tt)* }
            ),+ $(,)?]
        ),+ $(,)?]
    ) => {
        mod $module_name {
            mod wincode {
                $($(impl_sequence_sample_struct!(
                    derives = (wincode::SchemaWrite),
                    serializer = "wincode",
                    abi_digest = $abi_digest,
                    struct $struct_name { $($body)* }
                );)+)+
            }
            mod bincode {
                $($(impl_sequence_sample_struct!(
                    derives = (serde_derive::Serialize),
                    serializer = "bincode",
                    abi_digest = $abi_digest,
                    struct $struct_name { $($body)* }
                );)+)+
            }
        }
    };
}
```

File: frozen-abi/src/stable_abi/impls.rs [hunk 9]
```diff
@@ -398,7 +546,7 @@ mod tests {
     );
 
     const ABI_DIGEST_EQUIVALENT_KEY_VALUE_SEQUENCES: &str =
-        "9pGP5GGD2HxDRCQeDv3rPGTfVH9SzkZCXMSGHpZnzz4G";
+        "ATEDuwPHiE6J9TtzEuSvS7tMvtAR6pZuKFtzGioozixd";
```
Comment (puhtaytow) [inline frozen-abi/src/stable_abi/impls.rs]: the digest changed along with manual `with = ""` implementation, the previous one triggered clippy unsafe arithmetics

File: frozen-abi/src/stable_abi/impls.rs [hunk 10]
```diff
@@ -410,7 +558,7 @@ mod tests {
     )]
     struct TestEquivalentBTreeMapVsVecWincode {
         #[stable_abi_sample(
-            with = "(0..rng.random::<u16>() % 4).map(|i| (((i << 8) + rng.random::<u8>() as u16), rng.random())).collect()"
+            with = "(0..rng.random_range(0..4)).map(|i| (u16::from_be_bytes([i as u8, rng.random()]), rng.random())).collect()"
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: maybe let's do it as separate PR

Comment (puhtaytow) [reply frozen-abi/src/stable_abi/impls.rs]: https://github.com/anza-xyz/solana-sdk/pull/741

File: frozen-abi/src/stable_abi/impls.rs [hunk 11]
```diff
@@ -504,4 +652,148 @@ mod tests {
...
+                    #[stable_abi_sample(ctx = SequenceLenMax(1))]
+                    a: Vec<u8>,
+                },
+                struct UpToOneElementBTreeSet {
+                    #[stable_abi_sample(ctx = SequenceLenMax(1))]
+                    a: BTreeSet<u8>,
+                },
+                struct UpToOneElementHashMap {
+                    #[stable_abi_sample(ctx = SequenceLenMax(1))]
+                    a: HashMap<[u8; 0], u8>,
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: I think I prefer to flip the zero-sized array from keys to values, maybe better to actually use `()`, it's somewhat more obvious, since using `[u8; 0]` looks like a typo

and we should add (under separate digest) a test for actual non-zero key-value collecitons

File: frozen-abi/src/stable_abi/impls.rs [hunk 12]
```diff
@@ -504,4 +652,148 @@ mod tests {
...
+                    $($(impl_sequence_sample_struct!(
+                        derives = (wincode::SchemaWrite),
+                        serializer = "wincode",
+                        abi_digest = $abi_digest,
+                        struct $struct_name { $($body)* }
+                    );)+)+
+                }
+                mod bincode {
+                    use crate::stable_abi::context::SequenceLenMax;
+                    use std::collections::{HashSet, HashMap, VecDeque, BTreeSet, BTreeMap};
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: could move those imports to top level - `mod tests {` ?

**Comment by puhtaytow** [comment]
Should be good now 👍

File: frozen-abi/src/stable_abi/impls.rs [hunk 13]
```diff
@@ -504,4 +654,162 @@ mod tests {
...
+                a: HashMap<u8, u16>,
+            },
+            struct NonZeroKeyValueHashMap {
+                #[stable_abi_sample(ctx = SequenceLenMax(1))]
+                a: HashMap<u8, u16>,
+            },
+            struct NonZeroKeyValueBTreeMap {
+                #[stable_abi_sample(ctx = SequenceLenMax(1))]
+                a: BTreeMap<u8, u16>,
+            },
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: let's add `Vec<(u8, u16)>` too

File: frozen-abi/src/stable_abi/impls.rs [hunk 14]
```diff
@@ -504,4 +654,162 @@ mod tests {
...
+        derives = $derives:tt,
+        serializer = $abi_serializer:literal,
+        abi_digest = $abi_digest:expr,
+        struct $struct_name:ident { $($body:tt)* }
+    ) => {
+        #[derive $derives]
+        #[cfg_attr(
+            feature = "frozen-abi",
+            derive(
+                solana_frozen_abi_macro::StableAbi,
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: if we import StableAbi, StableAbitSample, frozen_abi, maybe inlining `impl_sequence_sample_struct` would also make sense

File: frozen-abi-macro/src/lib.rs [hunk 15]
```diff
@@ -148,6 +148,28 @@ struct StableAbiSampleOptions {
     ctx_expr: Option<TokenStream2>,
 }
 
+#[cfg(feature = "frozen-abi")]
+// `with = "...rng..."` expressions can come from inside `macro_rules!`; rewrite
+// `rng` to this proc macros call-site span so it resolves to the generated
+// `sample(..., rng)` parameter
+fn normalize_rng_ident(tokens: TokenStream2) -> TokenStream2 {
```
Comment (kskalski) [inline frozen-abi-macro/src/lib.rs]: a bit unfortunate we need this rewrite... is this because we nest macros or any rng being a macro parameter gets wrong binding?
actually, what does `rng` bind to or what is the error without this normalization?

**Comment by kskalski** [comment]
Another test we should have (another digest) is the `SequenceLenRange` with `min:1`

File: frozen-abi/src/stable_abi/impls.rs [hunk 16]
```diff
@@ -504,4 +654,166 @@ mod tests {
...
+                struct UpToOneElementHashSetDefault {
+                    a: HashSet<u8>,
+                },
+                struct UpToOneElementBTreeMap {
+                    #[stable_abi_sample(ctx = SequenceLenMax(1))]
+                    a: BTreeMap<u8, ()>,
+                },
+            ],
+            "Du8dTBApdeSxYTQVprkbMGvc5dLgCfUKKZ2z63qqr5Bj" => [
+            struct NonZeroVecTuple {
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: * move it below the "canonical" struct that has `with =`
* rename to UpToOneKeyValueVec (also rename other key value types to UpToOneKeyValue*)
* fix indentation

File: frozen-abi/src/stable_abi/impls.rs [hunk 17]
```diff
@@ -104,11 +111,154 @@ where
...
+
+impl_with_context_via! {
+    impl[T] for VecDeque<T>,
+    SequenceLenMax as ctx => SequenceLenRange { min: 0, max: ctx.0 },
+    where { T: StableAbi },
+}
+
+impl_with_context_via! {
+    impl[K, V] for BTreeMap<K, V>,
+    () as _ctx => SequenceLenRange { min: 0, max: 3 },
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: let's use consistent default max for all collections (except for hash-based of course) and best put it as a constant like `DEFAULT_COLLECTION_MAX_SAMPLE_LEN`

once we support `SequenceSortedInput` we could then match vec and btree with just that attribute

File: frozen-abi-macro/src/lib.rs (hunk 15 above)
Comment (puhtaytow) [reply frozen-abi-macro/src/lib.rs]: yes this is caused by macro nesting

```
error[E0425]: cannot find value `rng` in this scope
   --> frozen-abi/src/stable_abi/impls.rs:727:32
    |
727 |                         with = "(0..rng.random_range(0..=5)).map(|_| rng.random()).collect()"
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope
```
to my understanding this happens because generated scope thru macro rules is different than directly from type field sampling options

if you know how to fix it :rocket: :rocket:

Comment (puhtaytow) [reply frozen-abi-macro/src/lib.rs]: I might also try harder with clanker if thats blocker

Comment (kskalski) [reply frozen-abi-macro/src/lib.rs]: do we actually support `with = expr`? that would solve the issue

File: frozen-abi-macro/src/lib.rs [hunk 18]
```diff
@@ -166,7 +188,7 @@ fn parse_stable_abi_sample_options(field: &syn::Field) -> Result<StableAbiSample
                 let expr = value.parse::<Expr>().map_err(|err| {
                     Error::new(value.span(), format!("invalid `with` expression: {err}"))
                 })?;
-                with_expr = Some(quote! { #expr });
+                with_expr = Some(normalize_rng_ident(quote! { #expr }));
```
Comment (kskalski) [inline frozen-abi-macro/src/lib.rs]: Clankers for the rescue:

**Repro** (after removing `normalize_rng_ident`): tests fail with `cannot find value 'rng' in this scope` for every `with = "...rng..."` attribute that came out of a `macro_rules!` expansion (`MultiElementsWith`, `UpToOneElementWith`, `UpToOneKeyValueHashMapWith`, etc.).

**Why:** `LitStr::parse::<Expr>()` (in syn) tokenizes the string and then re-spans every resulting token to the `LitStr`'s own span. When the `LitStr` came from a `macro_rules!` body, that span carries the macro's *definition* hygiene context. The generated `fn sample(..., rng)` parameter is emitted via `quote!` with the proc-macro's `Span::call_site()` hygiene. Two different syntax contexts → name lookup fails.

The old `normalize_rng_ident` patched it by walking the token tree and rewriting only the `rng` ident to `Span::call_site()`. That's fragile (hard-codes one name, won't help any other identifier the user reaches for) and obscures what's actually wrong.

**Better fix** (frozen-abi-macro/src/lib.rs:186-195): swap `value.parse::<Expr>()` for `syn::parse_str::<Expr>(&value.value())`. `parse_str` goes through `proc_macro2::TokenStream::from_str`, which yields tokens at the proc macro's `Span::call_site()` — no re-spanning to the `LitStr`. Every identifier in the user's expression (including `rng`, `std::…`, anything else) now resolves in the same context as the generated parameter and the surrounding module.

All 94 frozen-abi tests pass, clippy is clean.


Let's do that as a separate PR, though it would be nice for the fix to actually tigger the error scenario.
Supposedly this is the minimum repro:
```rust
macro_rules! mk {
    ({ $($body:tt)* }) => {
        #[derive(StableAbi, StableAbiSample)]
        #[frozen_abi(abi_digest = "...", abi_serializer = "wincode")]
        struct S { $($body)* }
    };
}
mk!({ #[stable_abi_sample(with = "rng.random::<u8>()")] a: u8, });
```

File: frozen-abi/src/stable_abi/context.rs [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+/// (`Vec`, `VecDeque`, `HashMap`, `BTreeMap`) that bounds the sampled length
+/// to an inclusive range `min..=max`.
+///
+/// The collection's `random_with_context` draws a length uniformly from
+/// `min..=max` and produces that many elements. `min == max` pins the length
+/// exactly; `min == 0` means "up to `max` elements" (equivalent to
+/// `SequenceLenMax(max)`).
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub struct SequenceLenRange {
+    pub(super) min: usize,
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/context.rs]: it's a public struct that users are supposed to use, so those shoudl be public fields. we could also add a `const fn new(impl RangeBounds<usize>)` constructor

File: frozen-abi/src/stable_abi/impls.rs [hunk 20]
```diff
@@ -104,11 +113,154 @@ where
...
+{
+    fn random_with_context(rng: &mut (impl RngCore + ?Sized), ctx: SequenceLenRange) -> Self {
+        let len = rng.random_range(ctx.min..=ctx.max);
+        (0..len).map(|_| (K::random(rng), V::random(rng))).collect()
+    }
+}
+
+impl_with_context_via! {
+    impl[K, V, S] for HashMap<K, V, S>,
+    SequenceLenMax as ctx => SequenceLenRange { min: 0, max: ctx.0 },
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: how about adding `.into_range_context()` to `SequenceLenMax` ?

File: frozen-abi/src/stable_abi/context.rs [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+/// exactly; `min == 0` means "up to `max` elements" (equivalent to
+/// `SequenceLenMax(max)`).
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub struct SequenceLenRange {
+    pub min: usize,
+    pub max: usize,
+}
+
+impl SequenceLenRange {
+    pub const fn new(min: usize, max: usize) -> Self {
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/context.rs]: I think the range syntax would be nice to support, maybe just drop `const` - I guess we don't actually need it, since the we always evaluate the `ctx` or `convert` in body function, right?

Comment (puhtaytow) [reply frozen-abi/src/stable_abi/context.rs]: right. with const it wasnt possible, fixed now :+1:

File: frozen-abi/src/stable_abi/context.rs [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+    }
+}
+
+impl From<SequenceLenMax> for SequenceLenRange {
+    fn from(max: SequenceLenMax) -> Self {
+        Self::new(0..=max.0)
+    }
+}
+
+impl From<usize> for SequenceLenRange {
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/context.rs]: This kind of conversion is not really obvious or very useful, converting single number into range?

File: frozen-abi/src/stable_abi/impls.rs [hunk 23]
```diff
@@ -528,4 +679,164 @@ mod tests {
...
+            ],
+            "57dywdByMU7XcWbsfnXY5ChZdqne57zjJCn7uEjDKGNc" => [
+                struct UpToOneKeyValueHashMapWithRange {
+                    #[stable_abi_sample(
+                        with = "(0..rng.random_range(1..=1)).map(|_| (rng.random(), rng.random())).collect()"
+                    )]
+                    a: HashMap<u8, u16>,
+                },
+                struct UpToOneKeyValueVecRange {
+                    #[stable_abi_sample(ctx = SequenceLenRange{min: 1, max: 1})]
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: could this use the range syntax `SequenceLenRange::new(1..=1)` ?

File: frozen-abi/src/stable_abi/impls.rs [hunk 24]
```diff
@@ -104,11 +114,154 @@ where
...
+
+impl_with_context_via! {
+    impl[T] for VecDeque<T>,
+    SequenceLenMax as ctx => SequenceLenRange::from(ctx),
+    where { T: StableAbi },
+}
+
+impl_with_context_via! {
+    impl[K, V] for BTreeMap<K, V>,
+    () as _ctx => SequenceLenRange::from(DEFAULT_COLLECTION_MAX_SAMPLE_LEN),
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/impls.rs]: I would rather use the range syntax `SequenceLenRange::new(0..=DEFAULT_COLLECTION_MAX_SAMPLE_LEN)`

File: frozen-abi/src/stable_abi/context.rs (hunk 22 above)
Comment (puhtaytow) [reply frozen-abi/src/stable_abi/context.rs]: agreed :)

File: frozen-abi/src/stable_abi/context.rs [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+/// Context for `StableAbi<...>` impls on sequence-like collections
+/// that caps the sampled length at `0..=N`.
+///
+/// Equivalent to `SequenceLenRange { min: 0, max: N }`; the per-collection
+/// impl delegates to the `SequenceLenRange` impl.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct SequenceLenMax(pub usize);
+
+impl SequenceLenMax {
+    pub const fn new(max: usize) -> Self {
```
Comment (kskalski) [inline frozen-abi/src/stable_abi/context.rs]: this new seem redundant

**Review by kskalski — COMMENTED** [review]
LGTM, let's have some more eyes on that.

**Review by alexpyattaev — DISMISSED** [review]
LGTM

File: frozen-abi/src/stable_abi/impls.rs [hunk 26]
```diff
@@ -104,11 +114,154 @@ where
     }
 }
 
+// keep at most one element to mitigate iteration order differences
+impl_with_context_via! {
+    impl[K, V, S] for HashMap<K, V, S>,
```
Comment (joncinque) [inline frozen-abi/src/stable_abi/impls.rs]: Why does the macro use square brackets instead of angle brackets for the generic types? Wouldn't this be more expected for Rust users? Or is the idea for the macro to only be used internally, so we don't really care?

```suggestion
    impl<K, V, S> for HashMap<K, V, S>,
```

File: frozen-abi/src/stable_abi/impls.rs [hunk 27]
```diff
@@ -104,11 +114,154 @@ where
...
+    () as _ctx => SequenceLenRange::new(0..=DEFAULT_COLLECTION_MAX_SAMPLE_LEN_NON_DETERMINISTIC_ORDER),
```
Comment (joncinque) [inline frozen-abi/src/stable_abi/impls.rs]: The syntax here is a bit rough, and makes me wonder if we can use a closure instead of the macro and context type.

No need to change anything here, but would something like a closure potentially simplify this?

**Review by joncinque — COMMENTED** [review]
Just a small comment on the macro, looks great overall!

File: frozen-abi/src/stable_abi/impls.rs (hunk 27 above)
Comment (kskalski) [reply frozen-abi/src/stable_abi/impls.rs]: This macro might become public in the future, as it's quite a convenient helper, so definitely worth trying.

Seems like we can improve syntax quite a bit, let's try
```rust
macro_rules! impl_with_context_via {
    (
        impl<$($g:tt)*> StableAbi<$ctx:ty> for $self_ty:ty
        where { $($wc:tt)* },
        |$bind:ident| $convert:expr $(,)?
    ) => {
        impl<$($g)*> StableAbi<$ctx> for $self_ty
        where $($wc)*
        {
            fn random_with_context(
                rng: &mut (impl RngCore + ?Sized),
                $bind: $ctx,
            ) -> Self {
                Self::random_with_context(rng, $convert)
            }
        }
    };
}

impl_with_context_via! {
    impl<K, V, S> StableAbi<SequenceLenMax> for HashMap<K, V, S>
    where { K: StableAbi + Eq + Hash, V: StableAbi, S: BuildHasher + Default },
    |ctx| SequenceLenRange::from(ctx),
}

impl_with_context_via! {
    impl<T, S> StableAbi<()> for HashSet<T, S>
    where { T: StableAbi + Eq + Hash, S: BuildHasher + Default },
    |_| SequenceLenRange::new(0..=DEFAULT_COLLECTION_MAX_SAMPLE_LEN_NON_DETERMINISTIC_ORDER),
}
```

Comment (joncinque) [reply frozen-abi/src/stable_abi/impls.rs]: That's definitely a much nicer option already!

File: frozen-abi/src/stable_abi/impls.rs (hunk 26 above)
Comment (puhtaytow) [reply frozen-abi/src/stable_abi/impls.rs]: if im not mistaken it was due to local ambiguity when calling macro `impl_with_context_via`: multiple parsing options: built-in NTs tt ('g') 

but the new context_via syntax seem to work with <> now, thanks!

File: frozen-abi/src/stable_abi/impls.rs (hunk 27 above)
Comment (puhtaytow) [reply frozen-abi/src/stable_abi/impls.rs]: should be good now :+1: thank you guys :pray: :pray:

**Review by kskalski — APPROVED** [review]

**Review by joncinque — APPROVED** [review]
Looks great to me, thanks!
